### DISCUSSION TITLE
How jwk.Cache refresh works?

### DISCUSSION THREAD

**Discussion by ruspaul013** [lestrrat-go/jwx#1154] [Q&A]
Hello!

I am trying to understand how `jwk.WithMinRefreshInterval(15*time.Minute)` and others Refresh things works for JWKs. I added in my code this function that returns a jwk.Set and when I parse the keys I use `jwt.WithKeySet`, as was mention [here](https://github.com/lestrrat-go/jwx/discussions/957) .
```
type JWK struct {
	Cache jwk.Set
	URL   string
	lock  sync.Mutex
}

func (k *JWK) JwkURL(jwkURLPath string) error {
	log.Info().Interface("Using jwk URL", jwkURLPath)
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	jwkurl.lock.Lock()
	defer jwkurl.lock.Unlock()

	jwkCache := jwk.NewCache(ctx, jwk.WithRefreshWindow(10*time.Second))
	err := jwkCache.Register(jwkURLPath, jwk.WithMinRefreshInterval(10*time.Second), jwk.WithRefreshInterval(10*time.Second))
	if err != nil {
		fmt.Printf("failed to register jwk location: %s\n", err)
		return err
	}

	_, err = jwkCache.Refresh(ctx, jwkURLPath)
	if err != nil {
		fmt.Printf("failed to refresh JWKS: %s\n", err)
		return err
	}
	cached := jwk.NewCachedSet(jwkCache, jwkURLPath)

	k.Cache = cached
	k.URL = jwkURLPath
	return err
}

//Another function
func jwtTokenCheck(ctx context.Context, cached jwk.Set, fullMethod string, JWTAudienceRegexp string, JWTIsuuerRegexp string) (context.Context, SpiceDBClaims, error) {
...
       tokenInfo, err := jwt.ParseString(token, jwt.WithKeySet(cached))
...
}
``` 

I set the refresh intervals to 10s and I was expected to see requests made to the JKWs server almost every 10s to 20s approximately, but it seems that no request is made.  

Do I need to call Refresh by myself or something else should be added? 

Please let me know if you need more informations.

Thank you! :D

L.E: I forgot to mention that I initialize the cache at startup.

**Comment by lestrrat** [comment]
I haven't checked recently, but I think that was the intended behavior. Please file an issue with reproducible test case it it isn't working.

**Reply by ruspaul013** [reply]
Hello, I found what was the problem while trying to came up with a reproductible test case. Because I used a `context.WithCancel` in `jwk.NewCache`, it seems that the refresh is not happening. If I use first a `context.Background` an then one with cancel, as is used [here](https://huma.rocks/how-to/oauth2-jwt/#huma-auth-middleware), everything seems to work.
